docs/keps/20200619-federated-resource-status.md

```diff
@@ -0,0 +1,408 @@
+}
+
+```
+
+Nowadays `Conditions` hold the status of the federated actions (aka propagation status).
+In other words, it defines the conditions of the propagation status for a resource.
+
+```yaml
+- apiVersion: types.kubefed.io/v1beta1
+  kind: FederatedDeployment
+  metadata:
+    finalizers:
+    - kubefed.io/sync-controller
+    generation: 1
+    name: asystem
+    namespace: asystem
+    resourceVersion: "70174497"
+  spec:
+    placement:
+      clusters:
+      - name: cluster3
+      - name: cluster2
+      - name: cluster1
+    template:
+      metadata:
+        labels:
+          app: nginx
+      spec:
+        replicas: 3
+        selector:
+          matchLabels:
+            app: nginx
+        template:
+          metadata:
+            labels:
+              app: nginx
+          spec:
+            containers:
+            - image: nginx
+              name: nginx
+  status:
+    conditions:
+    - lastTransitionTime: "2020-05-25T19:47:59Z"
+      lastUpdateTime: "2020-05-25T19:47:59Z"
+      status: "True"
+      type: Propagation
+```
+
+`status.conditions` reports the latest status which defines the state of the propagation.
+Obviously it is not necessary to report all the clusters for which the propagation when
+successful.
+
+The intention in this proposal is to extend the current available `Conditions` to
+hold the status of the federated resources, e.g Ready, NotReady.
+
+The status of the federated resources should determine whether the resources satisfy
+a `Ready` condition, and otherwise report their error status.
+To do so, this property reports the status of the federated resources in their
+target clusters whenever a `ReadyCondition` is not satisfied.
+This condition would need to be identified per type or by the usage of an interface
+`IsReady` that determines this value per type of resource.
+By doing so, we ensure the `Conditions` property shows the status of only unhealthy
+resources.
+
+If we re-use the example from above and imagine a scenario where this `FederatedDeployment` resource remained `Ready=True` in two clusters, but crashed in `cluster3`.
+The value of `status.conditions` should reflect the new state for that specific cluster.
+
+```yaml
+- apiVersion: types.kubefed.io/v1beta1
+  kind: FederatedDeployment
+  metadata:
+    finalizers:
+    - kubefed.io/sync-controller
+    generation: 1
+    name: asystem
+    namespace: asystem
+    resourceVersion: "70174497"
+  spec:
+    placement:
+      clusters:
+      - name: cluster2
+      - name: cluster1
+    template:
+      metadata:
+        labels:
+          app: nginx
+      spec:
+        replicas: 3
+        selector:
+          matchLabels:
+            app: nginx
+        template:
+          metadata:
+            labels:
+              app: nginx
+          spec:
+            containers:
+            - image: nginx
+              name: nginx
+  status:
+    clusters:
+    - name: "cluster3"
+      status: "ReplicaFailure"
```

Comment:
I don't quite like the idea of coupling propagation status with object status on the same field.

For backwards compatibility, maybe still keep `clusters[x].status` as the propagation status, and introduce another `clusters[x].conditions` to capture the abnormal conditions for each clusters (e.g., `NotReady` type). In that way, we could also track other information like the message for the condition for users to triage problems from the mgmt cluster.